### PR TITLE
📊 Copernicus Climate Data API 

### PR DESCRIPTION
**Pull request by veronikasamborska1994**
hey @lucasrodes and @pabloarosado! I think this is now ready for review and you two seem like the best people to take a look! 

@lucasrodes do you mind taking a look at the following?

- [ ] snapshots/countries/2023-12-27/world_bank.zip.dvc
- [ ] snapshots/countries/2023-12-27/world_bank.py
- [ ] snapshots/climate/2023-12-20/surface_temperature.py
- [ ] snapshots/climate/2023-12-20/surface_temperature.nc.dvc
- [ ] pyproject.toml
- [ ] poetry.lock
- [ ] etl/steps/data/meadow/climate/2023-12-20/surface_temperature.py

@pabloarosado it'd be great if you could take a look at the garden/grapher steps for these

- [x] etl/steps/data/grapher/climate/2023-12-20/surface_temperature_monthly.py
- [x] etl/steps/data/grapher/climate/2023-12-20/surface_temperature_anomalies.py
- [x] etl/steps/data/grapher/climate/2023-12-20/surface_global_monthly_anomaly.py
- [x] etl/steps/data/garden/climate/2023-12-20/surface_temperature.py
- [x] etl/steps/data/garden/climate/2023-12-20/surface_temperature.meta.yml
- [x] etl/steps/data/garden/climate/2023-12-20/surface_temperature.excluded_countries.json
- [x] etl/steps/data/garden/climate/2023-12-20/surface_temperature.countries.json
- [x] dag/main.yml
- [x] dag/climate.yml

thank you both :)

### CODE REVIEW

**Comment by Marigold** [comment]
@veronikasamborska1994 me and @pabloarosado were trying to reduce size of `snapshots/climate/2023-12-20/surface_temperature.nc` (3.4gb is not that much, but given that we'll have more snapshots in the future, it's worth it). We found that gzipping the file (to `surface_temperature.nc.gz`) reduces its size from 3.4gb to 1.4gb. Is that feasible to do? (If decompression takes too long then it might not be worth it for example)

**Comment by veronikasamborska1994** [comment]
> @veronikasamborska1994 me and @pabloarosado were trying to reduce size of `snapshots/climate/2023-12-20/surface_temperature.nc` (3.4gb is not that much, but given that we'll have more snapshots in the future, it's worth it). We found that gzipping the file (to `surface_temperature.nc.gz`) reduces its size from 3.4gb to 1.4gb. Is that feasible to do? (If decompression takes too long then it might not be worth it for example)

hey @Marigold! thanks for the tip! I've now implemented your suggestion! decompression doesn't take too long - the meadow step is time-consuming for other reasons (ie. cropping the temperature data in longitude and latitude coordinates based on the shape of each country and averaging it). but anyway I don't think it's unreasonably long!

File: etl/steps/data/grapher/climate/2023-12-20/surface_global_monthly_anomaly.py
Comment (pabloarosado) [inline etl/steps/data/grapher/climate/2023-12-20/surface_global_monthly_anomaly.py:45]: Use `pivot` instead of `pivot_table`, it's safer, and it propagates metadata.

File: etl/steps/data/grapher/climate/2023-12-20/surface_global_monthly_anomaly.py
Comment (pabloarosado) [inline etl/steps/data/grapher/climate/2023-12-20/surface_global_monthly_anomaly.py:55]: You could define additional fields of metadata in the grapher yaml file instead.

File: etl/steps/data/grapher/climate/2023-12-20/surface_temperature_anomalies.py
Comment (pabloarosado) [inline etl/steps/data/grapher/climate/2023-12-20/surface_temperature_anomalies.py:27]: Use `pivot` instead of `pivot_table`.

Comment (pabloarosado) [reply etl/steps/data/grapher/climate/2023-12-20/surface_temperature_anomalies.py:27]: By the way, our `pivot` function has a `join_column_levels_with` which you may find useful, instead of having to rename indexes.

File: etl/steps/data/grapher/climate/2023-12-20/surface_temperature_anomalies.py
Comment (pabloarosado) [inline etl/steps/data/grapher/climate/2023-12-20/surface_temperature_anomalies.py:33]: Consider redefining metadata in the grapher yaml file, if needed.

File: etl/steps/data/grapher/climate/2023-12-20/surface_temperature_monthly.py
Comment (pabloarosado) [reply etl/steps/data/grapher/climate/2023-12-20/surface_temperature_monthly.py:23]: I think lines 23-27 can be replaced by simply:
```
df_pivot = tb[["country", "year", "month", "temperature_2m"]].pivot(index=["country", "month"], columns="year", values="temperature_2m", join_column_levels_with="_")
```

Comment (pabloarosado) [reply etl/steps/data/grapher/climate/2023-12-20/surface_temperature_monthly.py:23]: This will also automatically propagate metadata.

File: etl/steps/data/meadow/climate/2023-12-20/surface_temperature.py
Comment (pabloarosado) [inline etl/steps/data/meadow/climate/2023-12-20/surface_temperature.py:43]: If you are not using `da` anymore, you could rewrite `da = da - 273.15` to simplify and save some memory.

File: etl/steps/data/meadow/climate/2023-12-20/surface_temperature.py
Comment (pabloarosado) [inline etl/steps/data/meadow/climate/2023-12-20/surface_temperature.py:45]: I'd suggest moving the following lines (46-56) to the top of the run function, in the "Load inputs" part.

File: etl/steps/data/meadow/climate/2023-12-20/surface_temperature.py
Comment (pabloarosado) [inline etl/steps/data/meadow/climate/2023-12-20/surface_temperature.py:99]: You would need to manually change this on every update. Instead, you could do:
```
start_time = da_degc["time"].min().dt.date.astype(str).item()
end_time = da_degc["time"].max().dt.date.astype(str).item()
```

File: etl/steps/data/meadow/climate/2023-12-20/surface_temperature.py
Comment (pabloarosado) [inline etl/steps/data/meadow/climate/2023-12-20/surface_temperature.py:115]: Add `verify_integrity=True`.

**Review by pabloarosado — CHANGES_REQUESTED** [review]
Amazing work! I'm really excited to see the charts. Why don't you create some draft charts on the temporary staging server? You could also transfer them to production using `etl-staging-sync` (I can help you with that).
I found a few important things to change, which affect the data. Apart from that, I just added many small suggestions.
My main concern is that the processing of the raw data is quite tricky, it involves a reasonable level of expert knowledge to ensure there are no artifacts.
I think that, ideally, all this code should be on Copernicus' github, and their API should let us extract the averaged data by country. We could contact them directly to suggest that. If that's not feasible, at least it would be good to contact them, explain the processing, and ensure they agree with all of it. I can imagine that there are some additional issues we are not aware of, or small corrections that need to be applied, or caveats we should at least mention in the metadata.

File: snapshots/climate/2023-12-20/surface_temperature.nc.dvc
Comment (pabloarosado) [reply snapshots/climate/2023-12-20/surface_temperature.nc.dvc]: After compressing, I understand that now this is 1.4GB, which is a big improvement!

File: snapshots/climate/2023-12-20/surface_temperature.py
Comment (pabloarosado) [reply snapshots/climate/2023-12-20/surface_temperature.py:1]: I tried to repack the data and didn't find any way to make the file significantly smaller, so what you have (with Mojmir's suggestion of compressing) seems to be almost optimal.

**Comment by pabloarosado** [comment]
@lucasrodes I've actually reviewed all files, to have the full context, but still, feel free to have a look at the PR if you have some time.

File: etl/steps/data/garden/climate/2023-12-20/surface_temperature.meta.yml
Comment (veronikasamborska1994) [reply etl/steps/data/garden/climate/2023-12-20/surface_temperature.meta.yml:25]: can I call these shared fields in combination with additional descriptions that are specific to the indicator? they are not entirely identical

Comment (pabloarosado) [reply etl/steps/data/garden/climate/2023-12-20/surface_temperature.meta.yml:25]: You can try and see what happens.

File: etl/steps/data/garden/climate/2023-12-20/surface_temperature.meta.yml
Comment (veronikasamborska1994) [reply etl/steps/data/garden/climate/2023-12-20/surface_temperature.meta.yml:49]: yep, I've created it for plotting purposes grapher!

File: etl/steps/data/garden/climate/2023-12-20/surface_temperature.py
Comment (veronikasamborska1994) [reply etl/steps/data/garden/climate/2023-12-20/surface_temperature.py:36]: thanks! :)

File: etl/steps/data/meadow/climate/2023-12-20/surface_temperature.py
Comment (veronikasamborska1994) [reply etl/steps/data/meadow/climate/2023-12-20/surface_temperature.py:99]: goood point! thanks for noticing this!

File: etl/steps/data/garden/climate/2023-12-20/surface_temperature.meta.yml
Comment (veronikasamborska1994) [reply etl/steps/data/garden/climate/2023-12-20/surface_temperature.meta.yml:25]: I have tried it before (and just) and it doesn't work for me so if you know a solution then it'd be helpful if you shared it here.

File: etl/steps/data/garden/climate/2023-12-20/surface_temperature.py
Comment (veronikasamborska1994) [reply etl/steps/data/garden/climate/2023-12-20/surface_temperature.py:59]: we can pick any reference period (e.g., on their [website](https://climate.copernicus.eu/surface-air-temperature-november-2023) they display data for two different reference periods - 1981-2010 and 1991-2020). What I am inclined to do is just to take all the data available as a reference period as this is the most unbiased way of doing it. The original comment was wrong and I've now fixed it.

**Comment by veronikasamborska1994** [comment]
hey @pabloarosado! thanks for taking a look at the code so thoroughly and for all your suggestions - I have now made all of the changes that you've suggested with the exception of the partly repeated metadata field comment - I can't find a way of keeping some descriptions the same across different indicators but combining them with others. If you have a solution, please let me know but otherwise I am just going to keep it as it is. 

Also, here are some example charts we can create on the staging server:

- [ ] [Global yearly surface temperature anomalies](http://staging-site-copernicus-api/admin/charts/7524/edit)
- [ ] [Monthly surface temperatures since 1950](http://staging-site-copernicus-api/admin/charts/7523/edit)
- [ ] [Monthly surface temperature anomalies since 1950](http://staging-site-copernicus-api/admin/charts/7525/edit) - for this one I am guessing you have a similar chart from a different source so maybe we could compare the differences between your source's estimates and my calculations here? 
- [ ] [Global yearly temperature anomalies since 1950 by month](http://staging-site-copernicus-api/admin/charts/7526/edit) - not sure if we will end up using this one. 

Overall, I agree it'd be great if the source provided us with the country data but currently, they don't. I don't think the calculations that I am making here are particularly sophisticated (I am just doing an average and normalizing for varying grid cell sizes by using latitude-based weights calculated from the cosine of the latitude, as suggested in their  [tutorials](https://ecmwf-projects.github.io/copernicus-training-c3s/reanalysis-climatology.html)). But I will email them and explain the pipeline (as well as ask if they have something similar in place) as I agree it's definitely a good idea to get their thoughts on these calculations. I can imagine there will be caveats with reporting temperatures for smaller countries due to the spatial resolution of their data so it will be definitely good to know how big they think the issue with this would be. 

thanks again for taking a look! would be great if you could take a look at the changes and approve the pull request when you have some time.

**Comment by pabloarosado** [comment]
Hi @veronikasamborska1994 thanks for implementing all the changes. I haven't had a thorough second review, but the changes look good.

Just a quick note on how to define texts and then call them in different places:
```
definitions:
  common:
    unit: °C
    [...]
  temperature_anomaly: |-
    Temperature blah blah.
  something_else: |-
    Something else.

[...]

tables:
  surface_temperature:
    variables:
      temperature_2m:
        title: Average surface temperature
        [...]
        description_processing: |-
            - {definitions.temperature_anomaly}

            - {definitions.something_else}
```

I just tried this and it works. But anyway, you don't need to make these changes.

The only thing I'd suggest doing before merging is to add another normal grapher dataset with country-date data of temperatures. So, basically, a grapher dataset that imports the garden dataset as it is (and maybe it's also convenient to have one with the average yearly temperature). That would allow us to show a map chart, and press the play button to see how temperatures have been changing over time around the world. Do you agree? Apart from that, feel free to merge.

However, before publishing the charts I'd suggest:
1. Contacting the Copernicus team (I suppose you have already done so) to confirm that they don't provide country data anywhere, and that they agree with our processing and publication.
2. Checking with Hannah. For that, you can add a comment on [this GitHub issue](https://github.com/owid/owid-issues/issues/1280) briefly explaining your processing and assumptions (e.g. your reference temperature), add screenshots / links to your draft charts, and tag Hannah. You can also write on the data & research group and tag her, or do both. Maybe we could discuss during the next data & research meeting other possible charts we could create with this really interesting data.

Good work!

**Review by pabloarosado — APPROVED** [review]

**Comment by veronikasamborska1994** [comment]
> Hi @veronikasamborska1994 thanks for implementing all the changes. I haven't had a thorough second review, but the changes look good.
> 
> Just a quick note on how to define texts and then call them in different places:
> 
> ```
> definitions:
>   common:
>     unit: °C
>     [...]
>   temperature_anomaly: |-
>     Temperature blah blah.
>   something_else: |-
>     Something else.
> 
> [...]
> 
> tables:
>   surface_temperature:
>     variables:
>       temperature_2m:
>         title: Average surface temperature
>         [...]
>         description_processing: |-
>             - {definitions.temperature_anomaly}
> 
>             - {definitions.something_else}
> ```
> 
> I just tried this and it works. But anyway, you don't need to make these changes.
> 
> The only thing I'd suggest doing before merging is to add another normal grapher dataset with country-date data of temperatures. So, basically, a grapher dataset that imports the garden dataset as it is (and maybe it's also convenient to have one with the average yearly temperature). That would allow us to show a map chart, and press the play button to see how temperatures have been changing over time around the world. Do you agree? Apart from that, feel free to merge.
> 
> However, before publishing the charts I'd suggest:
> 
> 1. Contacting the Copernicus team (I suppose you have already done so) to confirm that they don't provide country data anywhere, and that they agree with our processing and publication.
> 2. Checking with Hannah. For that, you can add a comment on [this GitHub issue](https://github.com/owid/owid-issues/issues/1280) briefly explaining your processing and assumptions (e.g. your reference temperature), add screenshots / links to your draft charts, and tag Hannah. You can also write on the data & research group and tag her, or do both. Maybe we could discuss during the next data & research meeting other possible charts we could create with this really interesting data.
> 
> Good work!

thanks again @pabloarosado for taking such a thorough look at the code! :) and for the tip on the metadata!! 

I won't publish any charts until we either hear back from copernicus (I've contacted them but so far haven't heard anything back) as well as discuss these with Hannah when she is back! 

regarding this point, I agree it will be good to have this and will add a grapher step for these but I am not sure how good the maps will look like tbh given that the differences between countries are going to be more striking than the differences across years! I'll send you a link to some charts once these are done. 

> The only thing I'd suggest doing before merging is to add another normal grapher dataset with country-date data of temperatures. So, basically, a grapher dataset that imports the garden dataset as it is (and maybe it's also convenient to have one with the average yearly temperature). That would allow us to show a map chart, and press the play button to see how temperatures have been changing over time around the world. Do you agree? Apart from that, feel free to merge.

I've also added some decadal averages to the grapher monthly anomalies and temperatures datasets as [per discussion here ](https://owid.slack.com/archives/C46U9LXRR/p1704899853344569)

**Pull request merged**
